### PR TITLE
Recommend cpython agnosticism in cpp custom op tutorial

### PR DESCRIPTION
**Pull request by janeyx99**
## Description
In 2.6, PyTorch will better support python agnosticism in custom operators. This PR updates the tutorial to recommend the torch.ops.load_library path so that future custom op writers can follow this better pattern instead. These changes are paired with updates to our example extension-cpp repo that have now landed: https://github.com/pytorch/extension-cpp/pull/105

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

File: advanced_source/cpp_custom_ops.rst [hunk 1]
```diff
@@ -62,41 +62,30 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
...
   setup(name="extension_cpp",
         ext_modules=[
-            cpp_extension.CppExtension("extension_cpp", ["muladd.cpp"])],
-        cmdclass={'build_ext': cpp_extension.BuildExtension})
+            cpp_extension.CppExtension(
+              "extension_cpp",
+              ["muladd.cpp"]
+              py_limited_api=True)],
+        cmdclass={'build_ext': cpp_extension.BuildExtension},
+        options={"bdist_wheel": {"py_limited_api": "cp39"}}
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:70]: What does `{"py_limited_api": "cp39"}` mean? That 3.9 is the smallest version we support?

File: advanced_source/cpp_custom_ops.rst [hunk 2]
```diff
@@ -62,41 +62,30 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
...
-            cpp_extension.CppExtension(
-              "extension_cpp",
-              ["python_agnostic_code.cpp"],
-              py_limited_api=True)],
-        cmdclass={'build_ext': cpp_extension.BuildExtension},
-        options={"bdist_wheel": {"py_limited_api": "cp39"}}
-  )
-
-Note that you must specify ``py_limited_api=True`` both within ``setup``
+It is necessary to specify ``py_limited_api=True`` both within ``setup``
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:85]: nit: "as an argument to CppExtension/CUDAExtension"

File: advanced_source/cpp_custom_ops.rst (hunk 1 above)
Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:70]: Ah I see it's metnioned below

Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst:70]: yeh

Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst:70]: i can also add a comment

File: advanced_source/cpp_custom_ops.rst [hunk 3]
```diff
@@ -255,13 +244,41 @@ first load the C++ library that holds the custom operator definition
...
-2. If your C++ custom operator is located in a shared library object, you can
-   also use ``torch.ops.load_library("/path/to/library.so")`` to load it. This
-   is the blessed path for Python agnosticism, as you will not have a Python C
-   extension module to import. See `torchao __init__.py <https://github.com/pytorch/ao/blob/881e84b4398eddcea6fee4d911fc329a38b5cd69/torchao/__init__.py#L26-L28>`_
-   for an example.
+
+1. In this tutorial, our C++ custom operator is located in a shared library object,
+   and we use ``torch.ops.load_library("/path/to/library.so")`` to load it. This
+   is the blessed path for Python agnosticism, and you will not have a Python C
+   extension module to import. See our `extension_cpp/__init__.py <https://github.com/pytorch/extension-cpp/blob/e4c4eb822889ea67f191071fa627d750e04bf047/extension_cpp/__init__.py>`_
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:251]: I'm curious: can one still import a module with py_limited_api=True?

File: advanced_source/cpp_custom_ops.rst [hunk 4]
```diff
@@ -62,41 +62,30 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
...
 
 If you need to compile CUDA code (for example, ``.cu`` files), then instead use
 `torch.utils.cpp_extension.CUDAExtension <https://pytorch.org/docs/stable/cpp_extension.html#torch.utils.cpp_extension.CUDAExtension>`_.
 Please see `extension-cpp <https://github.com/pytorch/extension-cpp>`_ for an
 example for how this is set up.
 
-Starting with PyTorch 2.6, you can now build a single wheel for multiple CPython
-versions (similar to what you would do for pure python packages). In particular,
+Note that you can build a single wheel for multiple CPython versions (similar to
+what you would do for pure python packages) starting with PyTorch 2.6. Specifically,
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:79]: Write somewhere that this is known as "Python agnosticism", it's not clear

File: advanced_source/cpp_custom_ops.rst (hunk 3 above)
Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst:251]: yes lol

it worked fine with ao, but no promises on other variations of PYBIND11_MODULE

Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:251]: Defining [Py_LIMITED_API](https://docs.python.org/3/c-api/stable.html#c.Py_LIMITED_API) in the header should be sufficient to prevent the code from compiling if someone uses a non-py-limited-api.

So, I think we don't need to recommend torch.ops.load_library. Thoughts?

In general in open source I have preferred asking users to import c-extension modules over calling torch.ops.load_library, because there are various footguns associated with it

Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:251]: Also, if the pybind11_module is broken with Py_LIMITED_API, then there's also a standard cpython way of creating the extension module

File: advanced_source/cpp_custom_ops.rst (hunk 4 above)
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:79]:
```suggestion
what you would do for pure Python packages) starting with PyTorch 2.6. Specifically,
```

File: advanced_source/cpp_custom_ops.rst [hunk 5]
```diff
@@ -255,13 +244,41 @@ first load the C++ library that holds the custom operator definition
...
-   we created will load the C++ custom operator definitions.
-2. If your C++ custom operator is located in a shared library object, you can
-   also use ``torch.ops.load_library("/path/to/library.so")`` to load it. This
-   is the blessed path for Python agnosticism, as you will not have a Python C
-   extension module to import. See `torchao __init__.py <https://github.com/pytorch/ao/blob/881e84b4398eddcea6fee4d911fc329a38b5cd69/torchao/__init__.py#L26-L28>`_
-   for an example.
+
+1. In this tutorial, our C++ custom operator is located in a shared library object,
+   and we use ``torch.ops.load_library("/path/to/library.so")`` to load it. This
+   is the blessed path for Python agnosticism, and you will not have a Python C
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:250]: Thinking if blessed is the best word here =)
```suggestion
   is the recommended path for Python agnosticism, and you will not have a Python C
```

File: advanced_source/cpp_custom_ops.rst [hunk 6]
```diff
@@ -255,13 +244,41 @@ first load the C++ library that holds the custom operator definition
...
+  ), f"Expected one _C*.so file, found {len(so_files)}"
+  torch.ops.load_library(so_files[0])
+
+  from . import ops
+
+
+2. You may also see other custom extensions importing the Python C extension module.
+   The module would be created in C++ and then imported in Python, like the code below.
+   This code is not guaranteed to use the stable limited CPython API and would block
+   your extension from building a Python-agnostic wheel! AVOID the following:
```
Comment (svekars) [inline advanced_source/cpp_custom_ops.rst:271]: AVOID feels like you are screaming at me =)
```suggestion
   your extension from building a Python-agnostic wheel! Please avoid the following:
```

**Review by svekars — COMMENTED** [review]
Minor editorial nits.

File: advanced_source/cpp_custom_ops.rst (hunk 3 above)
Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst]: fixed it!

File: advanced_source/cpp_custom_ops.rst (hunk 6 above)
Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst]: hehe i am ;) but i've rephrased

File: advanced_source/cpp_custom_ops.rst (hunk 4 above)
Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst]: Changed it everywhere to CPython agnosticism

File: advanced_source/cpp_custom_ops.rst [hunk 7]
```diff
@@ -62,20 +62,64 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
...
+only using the `CPython Stable Limited API <https://docs.python.org/3/c-api/stable.html>`_,
+which is a requirement for the building a CPython agnostic wheel. If this requirement
+is not met, it is possible to build a wheel that looks CPython agnostic but will crash,
+or worse, be silently incorrect, in another CPython environment. Take care to avoid
+using unstable CPython APIs, for example APIs from libtorch_python (in particular
+pytorch/python bindings,) and to only use APIs from libtorch (ATen objects, operators
+and the dispatcher). We strongly recommend defining the ``Py_LIMITED_API`` flag to
+ensure the extension is compliant and safe as a CPython agnostic wheel.
+
+The second and third lines inform setuptools that you intend to build a CPython agnostic
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:102]: I'm lost, which second and third lines?

File: advanced_source/cpp_custom_ops.rst [hunk 8]
```diff
@@ -62,20 +62,64 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
...
+wheel and will influence the naming of the wheel accordingly. It is necessary to specify
+``py_limited_api=True`` as an argument to CppExtension/CUDAExtension and also as an option
+to the ``"bdist_wheel"`` command with the minimal supported CPython version (in this case,
+3.9):
+
+.. code-block:: python
+  setup(name="extension_cpp",
+        ext_modules=[
+            cpp_extension.CppExtension(
+              ...,
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:112]: Does this need to have the extra_compile_args={"cxx": ["-DPy_LIMITED_API=0x03090000"]}, too?

File: advanced_source/cpp_custom_ops.rst [hunk 9]
```diff
@@ -62,20 +62,64 @@ Using ``cpp_extension`` is as simple as writing the following ``setup.py``:
...
+              py_limited_api=True)],  # Build 1 wheel across multiple Python versions
+        ...,
+        options={"bdist_wheel": {"py_limited_api": "cp39"}}  # 3.9 is minimum supported Python version
+  )
+
+This ``setup`` would build one wheel that could be installed across multiple CPython
+versions ``>=3.9``.
```
Comment (zou3519) [inline advanced_source/cpp_custom_ops.rst:119]: Uh, does the wheel have the right platform tags to reject Python 3.8? Or is the advice to people to always use the smallest version that is not deprecated?

**Review by zou3519 — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]

File: advanced_source/cpp_custom_ops.rst (hunk 8 above)
Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst:112]: yes but i wanted to showcase the different lines (these were the 2nd and 3rd lines lol)

am gonna rephrase to try to sound better

File: advanced_source/cpp_custom_ops.rst (hunk 9 above)
Comment (janeyx99) [reply advanced_source/cpp_custom_ops.rst:119]: the latter hehe 

i haven't checked against 3.8, but it was cuz i chose 3.9 as the min version out there. will include a sentence for that

Comment (zou3519) [reply advanced_source/cpp_custom_ops.rst:119]: Yeah we need to tell people that the numbers they choose should be the min version pytorch supports

**Review by svekars — APPROVED** [review]
